[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Jibby2k1/SPS_Curriculum/blob/main/Intro_Programming/Intro_Python/Intro_Python.ipynb)


**Content Produced by UF Signal Processing Society**

**Authors: Raul Valle & Contributors**

# Intro to Python

Python is the lingua franca of scientific computing — nearly every other workshop in this curriculum assumes it. By the end you will write real programs, organize them with functions and classes, and use NumPy + Matplotlib the way the [DSP](../../Intro_DSP/README.md) and [ML](../../Intro_Mach_Learn/README.md) workshops expect.

## 0. Introduction

Why Python? It reads almost like pseudocode, it has a library for everything, and — as we saw in [Intro to GPU Systems](../../Intro_GPU/README.md) — its scientific libraries hand the heavy lifting to fast compiled code underneath. You write the *what*, the libraries handle the *how fast*.

## 1. Pre-requisites

None! This is the entry point of the programming track.

To run this notebook you need a Python environment with Jupyter. The easiest paths:

- **Google Colab** (zero install): upload this notebook at [colab.research.google.com](https://colab.research.google.com/).
- **Local**: install [miniconda](https://docs.conda.io/en/latest/miniconda.html), then
  `conda create -n sps python=3.12 numpy matplotlib jupyter` and select the `sps` kernel.

If you can run the cell below, you're ready.

In [1]:
print("Hello, SPS!")

Hello, SPS!


**What just happened.** The cell ran Python code and printed a string to the notebook's output area — if you saw `Hello, SPS!` appear below the cell, your environment (Colab or local Jupyter) is correctly wired up and every remaining cell in this notebook will run the same way: write code, run the cell, read the output directly underneath it.

---
### 🕐 Session 1 of 3 — *Language Core* (~35 min)
**Goal:** use variables, types, control flow, functions, and comprehensions.
**Feeds into:** Session 2 (data structures & objects).

---

<details><summary>🎓 <b>Teacher notes — Session 1: Language Core</b></summary>

**This is the entry point of the whole programming track — assume zero prior Python, but not zero prior programming.** Some students will have seen C, MATLAB, or another language; lean on that where it helps ("variables are labels stuck onto objects, not boxes with fixed types" is worth contrasting explicitly against C's declare-a-type-then-assign model if [Intro to C](../Intro_C.ipynb) came first for anyone in the room) and don't assume it where it doesn't (pure beginners need `print()`/variables from zero).

**The `/` vs `//` vs `%` trio is a small thing worth pausing on, because it's a real source of subtle bugs later.** `/` always returns a `float` in Python 3 (unlike C's integer division, where `7/2` is `3`) — a student porting logic from C to Python will get *silently* different, more precise results, which sounds harmless until it changes which branch of an `if` executes. `//` floors (not truncates) for negative numbers too — `-7 // 2` is `-4`, not `-3` — worth a quick mention if anyone asks.

**Control flow: sell "indentation is the syntax" as a feature that catches bugs early, not just a style choice.** A student who's used curly braces will initially find this restrictive; point out that inconsistent indentation is a compile-time error in Python, whereas mismatched braces in C can compile and silently do the wrong thing (the classic "if without braces" bug). The Collatz `while`-loop demo is a good moment to ask "will this loop definitely terminate?" — the Collatz conjecture is unproven in general, so strictly speaking this program's termination for *arbitrary* inputs is an open math problem, even though it visibly terminates for 27.

**Comprehensions are worth introducing as "the loop you'll actually write" rather than syntactic sugar.** They show up constantly in every downstream workshop (data filtering, list transforms before something becomes a NumPy array); the "for x in xs if cond(x)" reading in the intuition cell is the one to have students say out loud until it's automatic.
</details>

## 2. The Language Core

### 2.1. Variables & Types

Python variables are *labels stuck onto objects* — you don't declare types, the object itself knows what it is. The core types: `int`, `float`, `str`, `bool`.

In [2]:
x = 42          # int
y = 3.14        # float
name = "Fourier"  # str
ready = True    # bool

print(type(x), type(y), type(name), type(ready))

<class 'int'> <class 'float'> <class 'str'> <class 'bool'>


Arithmetic works as expected, with two divisions worth memorizing: `/` always gives a float, `//` floors to an integer, and `%` gives the remainder. `**` is exponentiation.

In [3]:
print(7 / 2)    # 3.5   true division
print(7 // 2)   # 3     floor division
print(7 % 2)    # 1     remainder
print(2 ** 10)  # 1024  power

3.5
3
1
1024


**What just happened.** `7 / 2` gave `3.5` (a `float`, always — even `4 / 2` would give `2.0`, not `2`), `7 // 2` floored that same division to `3`, and `7 % 2` gave the leftover `1` (and indeed `3*2 + 1 = 7`, so floor-division and modulo are consistent with each other by construction). `2 ** 10 = 1024` is ordinary exponentiation, with no separate `pow()` call needed. The two operators worth remembering by shape rather than name: `//` and `%` always satisfy `a == (a // b) * b + (a % b)` for any `a`, `b` — useful for sanity-checking either one if you forget which is which mid-derivation.

### 2.2. Strings

Strings are sequences: you can index them, slice them, and glue them together. *f-strings* (the `f"..."` form) are the standard way to format output.

In [4]:
signal = "sinusoid"
print(signal[0], signal[-1])   # first and last character
print(signal[:4])              # slice: characters 0..3
print(f"A {signal} has {len(signal)} letters.")

s d
sinu
A sinusoid has 8 letters.


**What just happened.** `signal[0]` and `signal[-1]` returned the first and last characters (`'s'` and `'d'`) — negative indices count backward from the end, so `-1` always means "the last one" regardless of the string's length. `signal[:4]` sliced the first four characters (`'sinu'`); Python slices are half-open, so `[:4]` means "up to but not including index 4," which is why it returns exactly 4 characters, not 5. The f-string interpolated both `signal` and `len(signal)` directly into the output string — `len("sinusoid")` is 8, matching the printed sentence.

### 2.3. Control Flow

Python uses **indentation** instead of braces — the whitespace *is* the syntax. If you took [Intro to C](../Intro_C.ipynb), everything here will feel familiar, just lighter.

In [5]:
for k in range(5):          # k = 0, 1, 2, 3, 4
    if k % 2 == 0:
        print(k, "even")
    else:
        print(k, "odd")

0 even
1 odd
2 even
3 odd
4 even


In [6]:
# while loops: collatz steps from 27
steps, m = 0, 27
while m != 1:
    m = m // 2 if m % 2 == 0 else 3 * m + 1
    steps += 1
print(f"27 reaches 1 in {steps} steps")

27 reaches 1 in 111 steps


**What just happened.** Starting from 27, the Collatz rule (halve if even, `3n+1` if odd) took 111 steps to reach 1 — and along the way it actually climbs *above* 27 several times (27 → 82 → 41 → 124 → ...) before eventually collapsing down, which is part of why the sequence looks unpredictable rather than monotonically shrinking. The `while m != 1` loop is a genuine leap of faith: it terminates for every starting value ever checked by computer (including numbers far larger than 27), but *no one has proven it terminates for all positive integers* — the Collatz conjecture is still open. This cell is a working example of a program whose termination the field cannot currently guarantee in general, even though it visibly halts right here.

### 2.4. Functions

Same idea as in every language: name a computation, give it inputs, return outputs. Default arguments and returning multiple values are everyday Python.

In [7]:
def db(power, reference=1.0):
    """Convert a power ratio to decibels."""
    import math
    return 10 * math.log10(power / reference)

print(db(100))        # 20 dB
print(db(2, 1))       # ~3 dB — the famous "3 dB point" from filter design

20.0
3.010299956639812


**What just happened.** `db(100)` used the default `reference=1.0` and returned exactly `20.0` — $10\log_{10}(100/1) = 10 \times 2 = 20$. `db(2, 1)` returned `3.010...`, the famous "3 dB point": a power ratio of 2 is *defined* as +3 dB (to two decimal places), which is why "half power" and "-3 dB" are used interchangeably everywhere in filter design — this function is computing the exact quantity that phrase refers to, not an approximation of it.

💡 **Intuition.** A *comprehension* is a for-loop folded into a single expression: `[f(x) for x in xs if cond(x)]` reads as "the list of f(x), for each x, keeping only those passing the condition." It's the idiom you'll see everywhere in scientific Python.

In [8]:
squares = [k**2 for k in range(10)]
even_squares = [s for s in squares if s % 2 == 0]
print(squares)
print(even_squares)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
[0, 4, 16, 36, 64]


**What just happened.** The first comprehension built all ten perfect squares from 0² through 9² in one expression, no explicit loop variable accumulation needed. The second comprehension then filtered *that* list down to the five even squares (`0, 4, 16, 36, 64`) — notice these are exactly the squares of the even inputs (`0, 2, 4, 6, 8`), since squaring preserves parity. Both comprehensions could equally be written as an explicit `for` loop with `.append()` calls; the comprehension is the same computation, just read left-to-right as "what to build, for what, filtered by what" instead of "start empty, loop, conditionally append."

---
### 🕐 Session 2 of 3 — *Data & Objects* (~35 min)
**Goal:** organize data with lists/dicts/sets; organize code with classes and modules.
**Builds on:** Session 1. &nbsp; **Feeds into:** Session 3 (scientific Python).

---

<details><summary>🎓 <b>Teacher notes — Session 2: Data & Objects</b></summary>

**The four-container table is the reference students will come back to for the rest of the curriculum — make sure the "why," not just the "what," lands.** The load-bearing distinction is mutability: a `tuple` can be a dict key or set member because it's hashable (immutable); a `list` cannot. This isn't pedantic trivia — it's exactly the rule that makes `point = (3.0, 4.0)` a reasonable dict key later but `samples` (a list) never can be. If a student tries `{[1,2]: "x"}` and gets a `TypeError: unhashable type`, that error *is* this lesson, live.

**The `dict`-as-phone-book intuition is worth reinforcing with the actual mechanism, at a level appropriate for this audience:** a dict computes a hash of the key and jumps straight to a bucket — no scanning — which is why lookup doesn't slow down as the dict grows, in contrast to a list's `in` check, which does scan linearly. This is the same $O(1)$-vs-$O(n)$ distinction the Data Structures in C workshop covers for trees and heaps, applied here to hashing instead.

**Classes: frame `self` as "the object doing the calling," not magic.** `s.duration()` is syntactic sugar for `Signal.duration(s)` — `self` is just the first parameter, filled in automatically. This demystifies both `self` and, later, why forgetting it in a method signature produces a confusing "missing argument" error. The `Signal` class here is intentionally minimal (just samples + sampling rate) — it's worth telling students this exact shape (data + the operations that make sense on it) is what they'll build repeatedly in the DSP workshops, just with more methods.

**Modules & imports:** the one habit worth instilling early is `from collections import Counter`-style targeted imports over `import *` — explicit imports make it obvious where a name came from when reading code later, which matters a lot once notebooks get long. If time allows, a one-line mention that `import` executes the module's top-level code exactly once (even if imported from multiple places) heads off a common later confusion about module-level state.
</details>

## 3. Data Structures

### 3.1. Lists, Tuples, Dicts, Sets

Four containers cover nearly everything:

| Container | Mutable? | Best for |
|---|---|---|
| `list` | yes | ordered collections you'll grow/change |
| `tuple` | no | fixed groups (coordinates, returns) |
| `dict` | yes | lookup by key |
| `set` | yes | membership tests, de-duplication |

In [9]:
samples = [0.1, 0.5, 0.2, 0.5]        # list
point = (3.0, 4.0)                     # tuple
freqs = {"A4": 440.0, "C5": 523.25}    # dict
unique = set(samples)                  # set: removes the duplicate 0.5

samples.append(0.9)
print(samples)
print(freqs["A4"])
print(unique)

[0.1, 0.5, 0.2, 0.5, 0.9]
440.0
{0.1, 0.5, 0.2}


**What just happened.** `samples.append(0.9)` mutated the list *in place* — the same `samples` object now has 5 elements, no reassignment needed, which is exactly what "lists are mutable" means concretely. `freqs["A4"]` jumped straight to `440.0` by key, not by scanning. And `set(samples)` — built from the *original* 4-element list before the append — silently dropped one of the two `0.5` entries, since sets only ever store one copy of each distinct value: `{0.1, 0.2, 0.5}` has 3 elements where the source list had 4. That silent de-duplication is the set's entire job, and also the reason to reach for one whenever "how many *distinct* values" is the real question.

💡 **Intuition.** A `dict` is a *phone book*: you don't scan for an entry, you jump straight to it by name. Lookup cost doesn't grow with size — that $O(1)$ behavior comes from hashing the key to a bucket index directly, the same trick behind hash tables in any language.

### 3.2. Classes

A class bundles *data* and the *functions that operate on it*. You've already used them — everything in Python is an object. `__init__` runs at construction; `self` is the object being operated on.

In [10]:
class Signal:
    """A minimal sampled-signal container."""
    def __init__(self, samples, fs):
        self.samples = list(samples)
        self.fs = fs                     # sampling rate in Hz

    def duration(self):
        return len(self.samples) / self.fs

    def scaled(self, gain):
        return Signal([gain * s for s in self.samples], self.fs)

s = Signal([0.0, 0.7, 1.0, 0.7, 0.0], fs=100)
print(f"{s.duration():.2f} s long")
print(s.scaled(2.0).samples)

0.05 s long
[0.0, 1.4, 2.0, 1.4, 0.0]


**What just happened.** `s.duration()` computed `len(self.samples) / self.fs = 5 / 100 = 0.05` seconds — the class turned "5 samples at 100 Hz" into the physically meaningful quantity (duration) without the caller needing to remember the formula. `s.scaled(2.0)` didn't modify `s` at all — it built and returned a *brand-new* `Signal` with every sample doubled (`[0.0, 1.4, 2.0, 1.4, 0.0]`), leaving the original untouched. That's a deliberate design choice worth naming: `scaled()` returns a new object rather than mutating `self.samples`, which is why you can chain `s.scaled(2.0).scaled(0.5)` safely without worrying about what state `s` itself ends up in.

### 3.3. Modules & Imports

Code is organized into *modules* (files) and *packages* (folders of files). `import` brings them in; the standard library is enormous.

In [11]:
import math
from collections import Counter

votes = ["fir", "iir", "fir", "fir", "iir"]
print(Counter(votes).most_common(1))
print(math.pi, math.e)

[('fir', 3)]
3.141592653589793 2.718281828459045


---
### 🕐 Session 3 of 3 — *Scientific Python* (~40 min)
**Goal:** compute with NumPy arrays and visualize with Matplotlib — the toolkit every later workshop assumes.
**Builds on:** Session 2. &nbsp; **Feeds into:** [DSP](../../Intro_DSP/README.md), [ML](../../Intro_Mach_Learn/README.md), [GPU](../../Intro_GPU/README.md) workshops.

---

<details><summary>🎓 <b>Teacher notes — Session 3: Scientific Python</b></summary>

**This session is the payoff for the whole workshop — nearly every other notebook in the curriculum assumes this toolkit. Treat it as the most important 40 minutes, not a wrap-up.** The "contiguous block of one type" framing in the intuition cell is the single fact that explains everything downstream: fixed-type, fixed-layout memory is what lets NumPy hand a whole array to compiled C/Fortran loops instead of looping in Python bytecode element by element.

**The vectorization timing demo is the session's centerpiece — walk expectations before running it.** The exact speedup number will vary by machine (the comment in the cell already flags this), but the *qualitative* result — a 10–50x-plus win from swapping a Python list comprehension for `2 * data + 1` — should hold everywhere. If a student's run shows a much smaller ratio, that's usually a sign the array is too small for the overhead of vectorized dispatch to pay off yet (worth demonstrating with `N=100` vs `N=1_000_000` if time allows) rather than a broken demo.

**Broadcasting is where most NumPy confusion concentrates later — spend real time on the rule, not just the demo.** "Shapes match from the right, or one side is 1" is the complete rule; walk `(3,4) - (4,)` explicitly as: align from the right, `4` matches `4`, and the missing leading dimension on the smaller array is treated as `1` and stretched to `3`. This is the mechanism, not magic — and it's worth previewing that broadcasting mistakes (accidentally subtracting along the wrong axis) are one of the most common silent-wrong-answer bugs in scientific Python, exactly like the `axis=0` vs `axis=1` choice in the column-mean demo.

**The spectrum plot is a deliberate preview, not a full lesson — resist the urge to explain the FFT here.** The goal is purely "here's a plot you'll make constantly," setting up Foundations of Signal Processing to do the actual explanation. If a student asks "wait, what *is* an FFT," the right answer in this session is "great question — next workshop," not a derailing tangent.
</details>

## 4. NumPy

💡 **Intuition.** A Python list is a bag of arbitrary objects; a NumPy array is a *contiguous block of numbers of one type* — the same "shelf" picture as C arrays. That layout is what lets NumPy hand whole-array operations to fast vectorized machine code, as demonstrated in [Intro to GPU Systems](../../Intro_GPU/Intro_GPU.ipynb) §2.2.

In [12]:
import numpy as np

t = np.linspace(0, 1, 8)   # 8 points from 0 to 1
x = np.sin(2 * np.pi * t)  # operates on ALL elements at once — no loop
print(t.round(3))
print(x.round(3))

[0.    0.143 0.286 0.429 0.571 0.714 0.857 1.   ]
[ 0.     0.782  0.975  0.434 -0.434 -0.975 -0.782 -0.   ]


### 4.1. Vectorization

The golden rule: **if you're writing a `for` loop over a NumPy array, look for the array operation instead.** It's shorter *and* orders of magnitude faster.

In [13]:
import time

N = 1_000_000
data = np.random.default_rng(42).random(N)

tic = time.perf_counter()
out_loop = [2 * v + 1 for v in data]           # Python loop
t_loop = time.perf_counter() - tic

tic = time.perf_counter()
out_vec = 2 * data + 1                          # vectorized
t_vec = time.perf_counter() - tic

print(f"loop:       {t_loop:.4f} s")
print(f"vectorized: {t_vec:.4f} s  ({t_loop / t_vec:.0f}x faster)")
# exact timings vary by machine; the ratio is the point

loop:       0.0821 s
vectorized: 0.0042 s  (20x faster)


### 4.2. Shapes, Slicing & Broadcasting

In [14]:
A = np.arange(12).reshape(3, 4)   # 3 rows, 4 columns
print(A)
print(A[1, :])      # row 1
print(A[:, 2])      # column 2
print(A.T.shape)    # transpose: (4, 3)

# broadcasting: subtract each column's mean from the column
print((A - A.mean(axis=0)).round(2))

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
[4 5 6 7]
[ 2  6 10]
(4, 3)
[[-4. -4. -4. -4.]
 [ 0.  0.  0.  0.]
 [ 4.  4.  4.  4.]]


**What just happened.** `A[1, :]` pulled out row 1 (`[4, 5, 6, 7]`) and `A[:, 2]` pulled out column 2 (`[2, 6, 10]`) — same array, two different slicing axes, and `A.T.shape` confirms the transpose swapped the shape from `(3, 4)` to `(4, 3)`. The broadcasting line is the one to slow down on: `A.mean(axis=0)` collapses the *row* axis, producing the length-4 vector of column means `[4, 5, 6, 7]` — and subtracting that length-4 vector from the `(3, 4)` matrix applies it to *every row*, giving `-4` across row 0, `0` across row 1 (which was already exactly the column means), and `+4` across row 2. No loop was written — NumPy stretched the `(4,)` vector against all 3 rows automatically because the trailing dimension matched.

💡 **Intuition.** *Broadcasting* stretches a smaller array across a bigger one without copying: subtracting a length-4 row of means from a (3, 4) matrix applies it to every row. If two shapes match from the right (or one side is 1), NumPy broadcasts.

## 5. Matplotlib

In [15]:
import matplotlib.pyplot as plt

fs = 500                                # sampling rate, Hz
t = np.arange(0, 1, 1 / fs)
clean = np.sin(2 * np.pi * 5 * t)       # 5 Hz sine
noisy = clean + 0.4 * np.random.default_rng(0).standard_normal(t.size)

plt.figure(figsize=(8, 3))
plt.plot(t, noisy, label="noisy", alpha=0.6)
plt.plot(t, clean, label="clean", linewidth=2)
plt.xlabel("time [s]"); plt.ylabel("amplitude")
plt.title("A 5 Hz sine in noise")
plt.legend(); plt.tight_layout(); plt.show()

/tmp/ipykernel_1833291/361073028.py:13: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.legend(); plt.tight_layout(); plt.show()


One more — the plot you will make a hundred times in this curriculum: a spectrum. (A preview of [Foundations of Signal Processing](../../Intro_DSP/Foundations_of_Signal_Processing_1.ipynb), where you'll learn what the FFT actually *is*.)

In [16]:
X = np.fft.rfft(noisy)
f = np.fft.rfftfreq(t.size, 1 / fs)

plt.figure(figsize=(8, 3))
plt.plot(f, np.abs(X) / t.size)
plt.xlim(0, 50)
plt.xlabel("frequency [Hz]"); plt.ylabel("|X(f)|")
plt.title("The 5 Hz peak pops right out of the noise")
plt.tight_layout(); plt.show()

/tmp/ipykernel_1833291/3806610988.py:9: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.tight_layout(); plt.show()


**What just happened.** `np.fft.rfft` computed the spectrum of the same noisy 5 Hz sine from the plot above, keeping only the non-negative frequencies (the "r" is for "real input" — for real-valued signals the negative-frequency half is a mirror image, so `rfft` skips computing it). `np.fft.rfftfreq` produced the matching frequency axis in Hz, so `plt.plot(f, np.abs(X)/t.size)` shows amplitude vs. real frequency directly. The result is a sharp spike right at 5 Hz rising clearly above a noisy floor — the noise is still there, spread thinly across every frequency, while the actual signal's energy is concentrated at the one frequency it belongs to. This is the visual argument for *why* spectral analysis is useful for noisy signals: a signal buried in time-domain noise can still stand out clearly once you look at the right axis.

## 6. Conclusion

You can now write Python programs, structure them with functions/classes, and lean on NumPy + Matplotlib for real numeric work. That's the exact toolkit the rest of the curriculum builds on.

---
## Where next

- [Foundations of Signal Processing](../../Intro_DSP/Foundations_of_Signal_Processing_1.ipynb) — that FFT peak, explained from first principles.
- [Intro to C](../Intro_C.ipynb) — what your Python objects look like one level down.
- [Intro to GPU Systems](../../Intro_GPU/README.md) — the same vectorization story, scaled to thousands of cores.
- [Intro to PyTorch](../../Intro_DL_4_Physics/intro_pytorch/intro_pytorch.ipynb) — NumPy's deep-learning sibling.